### PR TITLE
Remove hardcoded opaque.

### LINKED ISSUE #177: Need remove hardcode opaque.
https://github.com/intel/rust-spdm/blob/72c6977b964db4de9d0233643493d3bd73687e34/spdmlib/src/common/opaque.rs#L33-L107

### PR DESCRIPTION
**Pull request by longlongyang**
fix #177

### CODE REVIEW

**Comment by jyao1** [comment]
please test spdm 1.1 and 1.2.

File: spdmlib/src/common/opaque.rs [hunk 1]
```diff
@@ -30,81 +26,156 @@ pub const SUPPORTED_VERSION_LIST_SM_DATA_ID: u8 = 0x01;
...
+        }
+    }
+
+    fn read(r: &mut Reader) -> Option<Self> {
+        let buf = r.as_ref();
+        if buf.len() < 4 {
+            None
+        } else {
+            let id = u32::read_bytes(buf)?;
+            if id == DMTF_SPEC_ID {
```
Comment (jyao1) [inline spdmlib/src/common/opaque.rs]: I think we need to use SPDM version to determine.

SPDM 1.1 use GeneralOpaqueDataHeader.
SPDM 1.2+ use FM1OpaqueDataHeader

File: spdmlib/src/common/opaque.rs [hunk 2]
```diff
@@ -30,81 +26,156 @@ pub const SUPPORTED_VERSION_LIST_SM_DATA_ID: u8 = 0x01;
...
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SMOpaqueDataHeader {
+    General(GeneralOpaqueDataHeader),
+    FM1(FM1OpaqueDataHeader),
+}
+
+impl Codec for SMOpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        match self {
```
Comment (jyao1) [inline spdmlib/src/common/opaque.rs]: I am not sure how this work.

Does the caller need to know the difference?

I recommend to hide the info from the caller. The caller should just input SPDM version.
The rest should be handled by the callee.

**Comment by longlongyang** [comment]
> please test spdm 1.1 and 1.2.

spdm 1.2 is covered by CI

tested spdm 1.1 with below change

image

and below command:
```
cargo run -p spdm-responder-emu --no-default-features --features=spdm-ring,async-executor
cargo run -p spdm-requester-emu --no-default-features --features=spdm-ring,async-executor
```

test passed.

File: spdmlib/src/common/opaque.rs (hunk 1 above)
Comment (longlongyang) [reply spdmlib/src/common/opaque.rs]: > I think we need to use SPDM version to determine.
> 
> SPDM 1.1 use GeneralOpaqueDataHeader. SPDM 1.2+ use FM1OpaqueDataHeader

Suggestion applied

File: spdmlib/src/common/opaque.rs (hunk 2 above)
Comment (longlongyang) [reply spdmlib/src/common/opaque.rs]: > I am not sure how this work.
> 
> Does the caller need to know the difference?
> 
> I recommend to hide the info from the caller. The caller should just input SPDM version. The rest should be handled by the callee.

Suggestion applied

**Review by jyao1 — APPROVED** [review]

**Pull request merged**